File: Rust/Copilot/NBodySimulationKD.rs
use rayon::prelude::*;
use std::f64::consts::PI;

const G: f64 = 6.67430e-11; // Gravitational constant
const DT: f64 = 1e-3; // Time step
const NUM_BODIES: usize = 1000000; // Number of bodies
const THETA: f64 = 0.3; // Theta value for approximation

#[derive(Clone, Copy)]
struct Body {
    x: f64,
    y: f64,
    z: f64,
    vx: f64,
    vy: f64,
    vz: f64,
    mass: f64,
}

struct KDNode {
    body: Body,
    left: Option<Box<KDNode>>,
    right: Option<Box<KDNode>>,
    min: [f64; 3],
    max: [f64; 3],
}

fn initialize_bodies() -> Vec<Body> {
    let mut bodies = Vec::with_capacity(NUM_BODIES);
    bodies.push(Body {
        x: 0.0,
        y: 0.0,
        z: 0.0,
        vx: 0.0,
        vy: 0.0,
        vz: 0.0,
        mass: 1e30, // Central body mass
    });

    for i in 1..NUM_BODIES {
        let angle = 2.0 * PI * i as f64 / (NUM_BODIES - 1) as f64;
        bodies.push(Body {
            x: angle.cos() * 1e11,
            y: angle.sin() * 1e11,
            z: 0.0,
            vx: -angle.sin() * (G * bodies[0].mass / 1e11).sqrt(),
            vy: angle.cos() * (G * bodies[0].mass / 1e11).sqrt(),
            vz: 0.0,
            mass: 1e24, // Small body mass
        });
    }
    bodies
}

fn calculate_energy(bodies: &[Body]) -> f64 {
    bodies.par_iter().map(|body| {
        let kinetic = 0.5 * body.mass * (body.vx.powi(2) + body.vy.powi(2) + body.vz.powi(2));
        let potential: f64 = bodies.par_iter().filter(|&&other| body != &other).map(|other| {
            let dx = body.x - other.x;
            let dy = body.y - other.y;
            let dz = body.z - other.z;
            let distance = (dx.powi(2) + dy.powi(2) + dz.powi(2)).sqrt();
            -G * body.mass * other.mass / distance
        }).sum();
        kinetic + 0.5 * potential
    }).sum()
}

fn build_kdtree(bodies: &mut [Body], depth: usize) -> Option<Box<KDNode>> {
    if bodies.is_empty() {
        return None;
    }

    let axis = depth % 3;
    bodies.sort_by(|a, b| a[axis].partial_cmp(&b[axis]).unwrap());

    let median = bodies.len() / 2;
    let body = bodies[median];
    let left = build_kdtree(&mut bodies[..median], depth + 1);
    let right = build_kdtree(&mut bodies[median + 1..], depth + 1);

    let mut min = [body.x, body.y, body.z];
    let mut max = [body.x, body.y, body.z];

    if let Some(ref left_node) = left {
        for i in 0..3 {
            min[i] = min[i].min(left_node.min[i]);
            max[i] = max[i].max(left_node.max[i]);
        }
    }

    if let Some(ref right_node) = right {
        for i in 0..3 {
            min[i] = min[i].min(right_node.min[i]);
            max[i] = max[i].max(right_node.max[i]);
        }
    }

    Some(Box::new(KDNode { body, left, right, min, max }))
}

fn calculate_force(node: &KDNode, body: &Body, ax: &mut f64, ay: &mut f64, az: &mut f64) {
    let dx = node.body.x - body.x;
    let dy = node.body.y - body.y;
    let dz = node.body.z - body.z;
    let distance = (dx.powi(2) + dy.powi(2) + dz.powi(2)).sqrt();

    let size = node.max.iter().zip(node.min.iter()).map(|(max, min)| max - min).fold(0.0, f64::max);

    if size / distance < THETA || (node.left.is_none() && node.right.is_none()) {
        let force = G * node.body.mass / distance.powi(3);
        *ax += force * dx;
        *ay += force * dy;
        *az += force * dz;
    } else {
        if let Some(ref left) = node.left {
            calculate_force(left, body, ax, ay, az);
        }
        if let Some(ref right) = node.right {
            calculate_force(right, body, ax, ay, az);
        }
    }
}

fn kick_step(bodies: &mut [Body], root: &KDNode) {
    let accelerations: Vec<(f64, f64, f64)> = bodies.par_iter().map(|body| {
        let mut ax = 0.0;
        let mut ay = 0.0;
        let mut az = 0.0;
        calculate_force(root, body, &mut ax, &mut ay, &mut az);
        (ax, ay, az)
    }).collect();

    bodies.par_iter_mut().zip(accelerations.par_iter()).for_each(|(body, &(ax, ay, az))| {
        body.vx += ax * DT;
        body.vy += ay * DT;
        body.vz += az * DT;
    });

    bodies.par_iter_mut().for_each(|body| {
        body.x += body.vx * DT;
        body.y += body.vy * DT;
        body.z += body.vz * DT;
    });
}

fn main() {
    let mut bodies = initialize_bodies();

    let initial_energy = calculate_energy(&bodies);
    println!("Initial energy: {}", initial_energy);

    for _ in 0..1000 {
        let root = build_kdtree(&mut bodies, 0).unwrap();
        kick_step(&mut bodies, &root);
    }

    let final_energy = calculate_energy(&bodies);
    println!("Final energy: {}", final_energy);
}

File: Julia/Copilot/NBodySimulationKD.jl
using Base.Threads

const G = 6.67430e-11 # Gravitational constant
const DT = 1e-3 # Time step
const NUM_BODIES = 1000000 # Number of bodies
const THETA = 0.3 # Theta value for approximation

struct Body
    x::Float64
    y::Float64
    z::Float64
    vx::Float64
    vy::Float64
    vz::Float64
    mass::Float64
end

struct KDNode
    body::Body
    left::Union{KDNode, Nothing}
    right::Union{KDNode, Nothing}
    min::Vector{Float64}
    max::Vector{Float64}
end

function initialize_bodies()
    bodies = Vector{Body}(undef, NUM_BODIES)
    bodies[1] = Body(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1e30) # Central body mass

    for i in 2:NUM_BODIES
        angle = 2 * π * (i - 1) / (NUM_BODIES - 1)
        bodies[i] = Body(
            cos(angle) * 1e11,
            sin(angle) * 1e11,
            0.0,
            -sin(angle) * sqrt(G * bodies[1].mass / 1e11),
            cos(angle) * sqrt(G * bodies[1].mass / 1e11),
            0.0,
            1e24 # Small body mass
        )
    end
    return bodies
end

function calculate_energy(bodies)
    energy = 0.0
    @threads for i in 1:length(bodies)
        body = bodies[i]
        kinetic = 0.5 * body.mass * (body.vx^2 + body.vy^2 + body.vz^2)
        potential = 0.0
        for j in 1:length(bodies)
            other = bodies[j]
            if body != other
                dx = body.x - other.x
                dy = body.y - other.y
                dz = body.z - other.z
                distance = sqrt(dx^2 + dy^2 + dz^2)
                potential -= G * body.mass * other.mass / distance
            end
        end
        @atomic energy += kinetic + 0.5 * potential
    end
    return energy
end

function build_kdtree(bodies, depth=0)
    if isempty(bodies)
        return nothing
    end

    axis = depth % 3 + 1
    sort!(bodies, by=b -> getfield(b, (:x, :y, :z)[axis]))

    median = length(bodies) ÷ 2
    node = KDNode(
        bodies[median],
        build_kdtree(bodies[1:median-1], depth + 1),
        build_kdtree(bodies[median+1:end], depth + 1),
        [bodies[median].x, bodies[median].y, bodies[median].z],
        [bodies[median].x, bodies[median].y, bodies[median].z]
    )

    for i in 1:3
        if node.left != nothing
            node.min[i] = min(node.min[i], node.left.min[i])
            node.max[i] = max(node.max[i], node.left.max[i])
        end
        if node.right != nothing
            node.min[i] = min(node.min[i], node.right.min[i])
            node.max[i] = max(node.max[i], node.right.max[i])
        end
    end

    return node
end

function calculate_force(node, body, ax, ay, az)
    if node == nothing || node.body == nothing
        return
    end

    dx = node.body.x - body.x
    dy = node.body.y - body.y
    dz = node.body.z - body.z
    distance = sqrt(dx^2 + dy^2 + dz^2)

    size = max(node.max[1] - node.min[1], max(node.max[2] - node.min[2], node.max[3] - node.min[3]))

    if size / distance < THETA || (node.left == nothing && node.right == nothing)
        force = G * node.body.mass / (distance^3)
        ax[] += force * dx
        ay[] += force * dy
        az[] += force * dz
    else
        calculate_force(node.left, body, ax, ay, az)
        calculate_force(node.right, body, ax, ay, az)
    end
end

function kick_step!(bodies, root)
    @threads for i in 1:length(bodies)
        body = bodies[i]
        ax, ay, az = Ref(0.0), Ref(0.0), Ref(0.0)
        calculate_force(root, body, ax, ay, az)
        body.vx += ax[] * DT
        body.vy += ay[] * DT
        body.vz += az[] * DT
    end

    @threads for i in 1:length(bodies)
        body = bodies[i]
        body.x += body.vx * DT
        body.y += body.vy * DT
        body.z += body.vz * DT
    end
end

function main()
    bodies = initialize_bodies()

    initial_energy = calculate_energy(bodies)
    println("Initial energy: $initial_energy")

    for step in 1:1000
        root = build_kdtree(bodies)
        kick_step!(bodies, root)
    end

    final_energy = calculate_energy(bodies)
    println("Final energy: $final_energy")
end

main()
